lib/modules/manager/cargo/extract.ts

```
@@ -234,7 +236,23 @@ export async function extractPackageFile(
   const res: PackageFileContent = { deps };
   // istanbul ignore if
   if (lockFileName) {
+    logger.debug(
+      `Found lock file ${lockFileName} for packageFile: ${packageFile}`
+    );
     res.lockFiles = [lockFileName];
+
+    const versionsByPackage = await extractLockFileVersions(lockFileName);
+    for (const dep of deps) {
+      const versions = versionsByPackage.get(`${dep.depName!}`) ?? [];
+      const lockedVersion = cargoVersioning.getSatisfyingVersion(
```

should have a try catch, this function can throw on invalid versions.